Source/deviceinfo/DeviceInfo.cpp

```diff
@@ -611,106 +614,574 @@ class DeviceInfoLink : public WPEFramework::RPC::SmartInterfaceType<WPEFramework
 
 };
 
+class MemoryCachedDeviceInfo
+{
+public:
+    MemoryCachedDeviceInfo& operator=(const MemoryCachedDeviceInfo&) = delete;
+    MemoryCachedDeviceInfo(const MemoryCachedDeviceInfo&) = delete;
+    static MemoryCachedDeviceInfo& Instance()
+    {
+        return WPEFramework::Core::SingletonType<MemoryCachedDeviceInfo>::Instance();
+    }
+private:
+    MemoryCachedDeviceInfo():
+        _hdr_atmos_cec(0),
+        _outputResolution(),
+        _audioOutput(),
+        _videoOutput(),
+        _deviceInfoLink(DeviceInfoLink::Instance())
+    {
+        _architecture.Clear();
+        _chipsetName.Clear();
+        _modelName.Clear();
+        _modelYear.Clear();
+        _systemIntegraterName.Clear();
+        _friendlyName.Clear();
+        _platformName.Clear();
+        _firmwareVersion.Clear();
+        _idStr.Clear();
+        _supportedHdcp.Clear();
+        _maxOutputResolution.Clear();
+    }
+
+    ~MemoryCachedDeviceInfo()
+    {
+    }
+
+public:
+    uint32_t Deviceinfo_architecure(char buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_architecture.IsSet())
+        {
+            strncpy(buffer, _architecture.Value().c_str(), _architecture.Value().length());
+            *length = _architecture.Value().length();
+        }
+        else
+        {
+            result = _deviceInfoLink.Deviceinfo_architecure(buffer,length);
+            if(result == Core::ERROR_NONE)
+            {
+                _architecture = buffer;
+            }
+        }
+        return result;
+
+    }
+
+    uint32_t Deviceinfo_chipset(char buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_chipsetName.IsSet())
+        {
+            strncpy(buffer, _chipsetName.Value().c_str(), _chipsetName.Value().length());
+            *length = _chipsetName.Value().length();
+        }
+        else
+        {
+            result = _deviceInfoLink.Deviceinfo_chipset(buffer,length);
+            if(result == Core::ERROR_NONE)
+            {
+                _chipsetName = buffer;
+            }
+        }
+        return result;
+
+    }
+
+    uint32_t Deviceinfo_firmware_version(char buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_firmwareVersion.IsSet())
+        {
+            strncpy(buffer, _firmwareVersion.Value().c_str(), _firmwareVersion.Value().length());
+            *length = _firmwareVersion.Value().length();
+        }
+        else
+        {
+            result = _deviceInfoLink.Deviceinfo_firmware_version(buffer,length);
+            if(result == Core::ERROR_NONE)
+            {
+                _firmwareVersion = buffer;
+            }
+        }
+        return result;
+
+    }
+
+    uint32_t Deviceinfo_id(uint8_t buffer[], uint8_t* length)
+    {
+        uint32_t result = Core::ERROR_NONE;
+        if(_id.size()>0)
+        {
+            std::copy(_id.begin(), _id.end(), buffer);
```

Comment:
This could fail if *length was not big enough to hold the previously returned id array